lib/Sema/Constraints.cpp

```diff
@@ -0,0 +1,464 @@
+{
+    assert(
+        kind == ConstraintKind::ValueMember
+        || kind == ConstraintKind::UnresolvedValueMember
+    );
+    assert(member && "Member constraint has no member");
+    assert(locator && "Member constraint has no locator");
+
+    std::copy(
+        typeVars.begin(), typeVars.end(), getTypeVariablesBuffer().begin()
+    );
+}
+
+Constraint::Constraint(
+    glu::ast::ASTNode node, bool isDiscarded, glu::ast::ASTNode *locator,
+    llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+)
+    : _kind(ConstraintKind::SyntacticElement)
+    , _numTypeVariables(typeVars.size())
+    , _hasFix(false)
+    , _hasRestriction(false)
+    , _isActive(false)
+    , _isDisabled(false)
+    , _rememberChoice(false)
+    , _isFavored(false)
+    , _isDiscarded(isDiscarded)
+    , _syntacticElement { node }
+    , _locator(locator)
+{
+    std::copy(
+        typeVars.begin(), typeVars.end(), getTypeVariablesBuffer().begin()
+    );
+}
+
+Constraint::Constraint(
+    glu::types::Ty type, glu::ast::FunctionDecl *choice,
+    glu::ast::ASTNode *locator,
+    llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+)
+    : _kind(ConstraintKind::BindOverload)
+    , _numTypeVariables(typeVars.size())
+    , _hasRestriction(false)
+    , _isActive(false)
+    , _rememberChoice(false)
+    , _isFavored(false)
+    , _overload { type }
+    , _locator(locator)
+{
+    std::copy(
+        typeVars.begin(), typeVars.end(), getTypeVariablesBuffer().begin()
+    );
+    *getTrailingObjects<glu::ast::FunctionDecl *>() = choice;
+}
+
+/// Recursively gather the set of type variables referenced by this constraint.
+static void gatherReferencedTypeVars(
+    Constraint *constraint,
+    llvm::SmallPtrSetImpl<glu::types::TypeVariableTy *> &typeVars
+)
+{
+    switch (constraint->getKind()) {
+    case ConstraintKind::Disjunction:
+        for (auto nested : constraint->getNestedConstraints())
+            gatherReferencedTypeVars(nested, typeVars);
+        return;
+
+    case ConstraintKind::Conjunction:
+        typeVars.insert(
+            constraint->getTypeVariables().begin(),
+            constraint->getTypeVariables().end()
+        );
+        return;
+    case ConstraintKind::Bind:
+    case ConstraintKind::BindToPointerType:
+    case ConstraintKind::ArgumentConversion:
+    case ConstraintKind::Conversion:
+    case ConstraintKind::OperatorArgumentConversion:
+    case ConstraintKind::CheckedCast:
+    case ConstraintKind::Equal:
+    case ConstraintKind::UnresolvedValueMember:
+    case ConstraintKind::ValueMember:
+    case ConstraintKind::Defaultable:
+    case ConstraintKind::FallbackType:
+    case ConstraintKind::ExplicitGenericArguments:
+    case ConstraintKind::LValueObject:
+        typeVars.insert(
+            llvm::cast<glu::types::TypeVariableTy>(constraint->getFirstType())
+        );
+        typeVars.insert(
+            llvm::cast<glu::types::TypeVariableTy>(constraint->getSecondType())
+        );
```

Comment:
llvm::cast will crash if the cast fails... and there's no point in constraints if everything's a type variable, so of course it will fail